### PR TITLE
Refactor and improve MonologueProcess

### LINKED ISSUE #15: Monologue crashes during Test if ActionOption has invalid option ID
**Version:** Tested in v2.1.2, still present in v2.2.3
**Severity:** Low

# Steps to Reproduce
1. Create an ActionNode.
2. Select ActionOption as the action type.
3. Enter some option ID that doesn't exist, the panel will tell you that it doesn't exist.
4. Start the Test anyway.

# Current Behavior
Monologue will crash when it reaches this ActionNode. Changes are saved so it's not a big deal, but it's an inconvenience.

# Expected Behavior
The node should be skipped with a toast notification saying option ID was not found. The user should be informed of this error. 

### Photoshopped Preview

missing_notification

### PR DESCRIPTION
**Pull request by RailKill**
This PR cleans up and separates functions a little more clearly in MonologueProcess in preparation for future adaptations to other engines/languages as part of discussion #24. Having well-defined, atomic operations in its own function help with future extensibility, adaptability and testability, makes it more API-like.

# Features
- Improved conditional text processing in NodeSentence.
    
    **Previous Behavior**
    You can have `{{ if my_variable then something else another }}` in NodeSentence, and if `my_variable` is true, `something` will be emitted in `monologue_sentence()`. Only boolean variables were supported.

    **New Behavior**
    You can now nest and evaluate expressions in the conditional text. See `Test/Unit/TestMonologueProcess.gd` for automated test examples. Something like:
    ```
    {{ if variable_one + 50 < variable_two then if variable_three then a else 10 + 2 else "variable_one" }}
    ```
    will work. Any words in curly braces will be substituted with variable values if they have the same name unless they are surrounded by quotation marks which indicate they are Strings to be parsed as-is. If they do not exist as variables, you don't need quotation marks.

- Support runtime loading of .wav and .ogg audio files in `SfxLoader`. Previously, only .mp3 is supported.
- `Notification.gd`'s `print()` now excludes color codes and BB tags, making console logs cleaner and easier to read.

# Bug Fixes
- Fix #15

# Developer Notes
- `SfxPlayer.gd` has been removed to streamline things. `SfxLoader` and `AudioStreamPlayer2D` is sufficient.
- The `kalimba.wav` and `mystery_sting.ogg` for unit test were recorded by me. I played the instruments, no copyright issues.

## Summary by Sourcery

Refactor MonologueProcess for better clarity and future extensibility, enhance conditional text processing, and add support for additional audio formats. Improve console log readability and streamline audio handling by removing SfxPlayer.gd. Update build configurations and CI workflow, and introduce extensive unit tests for MonologueProcess.

New Features:
- Enhance conditional text processing in NodeSentence to support nested and evaluated expressions, allowing more complex logic in dialogue scripts.
- Add support for runtime loading of .wav and .ogg audio files in SfxLoader, expanding the range of audio formats that can be used.

Bug Fixes:
- Resolve issue #15, addressing a specific bug in the MonologueProcess.

Enhancements:
- Refactor MonologueProcess to improve function separation and clarity, facilitating future adaptations and improving extensibility and testability.
- Remove SfxPlayer.gd to streamline audio handling, as SfxLoader and AudioStreamPlayer2D are now sufficient.
- Improve console log readability by removing color codes and BB tags from Notification.gd.

Build:
- Update export_presets.cfg to exclude certain directories and adjust texture format settings for different platforms.

CI:
- Modify GitHub Actions workflow to update the path for unit tests, ensuring correct test execution.

Tests:
- Introduce comprehensive unit tests for MonologueProcess, covering various functionalities such as conditional text processing, audio playback, and node processing.

### CODE REVIEW

File: Test/Scripts/MonologueProcess.gd [hunk 1]
```diff
@@ -2,368 +2,410 @@ class_name MonologueProcess
...
-		return text
+	for result in results:
+		var target = result.get_string()
+		var replacement = str(evaluate_expression(target))
+		if target != replacement:
+			text = text.replace("{{" + target + "}}", replacement)
+	return text
+
+
+func evaluate_expression(expression: String) -> Variant:
```
[the remark this answers is not in the document]
Comment (RailKill) [reply Test/Scripts/MonologueProcess.gd:254]: this application uses a custom "if-then-else" syntax that isn't the same as the Godot engine, so a parsing library is not an option. the custom syntax is more natural for the user, and performance shouldn't be too much of a concern as the use case for this evaluation of expressions is a few lines at most

File: Test/Scripts/SfxLoader.gd [hunk 2]
```diff
@@ -1,30 +1,96 @@
...
+	return OK
 
 
-func clear():
-	for child in get_children():
-		child.queue_free()
+## Create an AudioStreamWAV based on given file bytes. This code is adapted
+## from [url]https://github.com/Gianclgar/GDScriptAudioImport[/url] which is
+## MIT-licensed. 24-bit WAV is not converted, but rejected.
+func parse_wav_data(bytes: PackedByteArray) -> AudioStreamWAV:
```
[the remark this answers is not in the document]
Comment (RailKill) [reply Test/Scripts/SfxLoader.gd:56]: it's just parsing the WAV header, i included a URL in the comments so it should suffice. i don't want to include too much inline comments, better to refer to an external reference. WAV is a common format, better to read the byte header format online

File: Test/Scripts/MonologueProcess.gd [hunk 3]
```diff
@@ -2,368 +2,410 @@ class_name MonologueProcess
...
+
+func find_node_from_id(node_id: Variant) -> Dictionary:
+	if node_id is String:
+		for node in node_list:
+			if node.get("ID") == node_id:
+				return node
+	return {}
+
+
+func process_node(node: Dictionary) -> void:
```
[the remark this answers is not in the document]
Comment (RailKill) [reply Test/Scripts/MonologueProcess.gd:100]: yeah probably in the future, i think it's still ok for now

File: Test/Scripts/MonologueProcess.gd [hunk 4]
```diff
@@ -2,368 +2,410 @@ class_name MonologueProcess
...
-		var variable = get_variable(var_name)
+	var parser = Expression.new()
+	var result = null
+	if expression:
+		parser.parse(expression)
+		result = parser.execute([], parser, false)
+	return expression if parser.has_execute_failed() or not result else result
+
+
+func substitute_variables(expression: String) -> String:
```
[the remark this answers is not in the document]
Comment (RailKill) [reply Test/Scripts/MonologueProcess.gd:287]: good suggestion, latest commit now use the string builder pattern

**Comment by atomic-junky** [comment]
I'll test that as soon as I can but it looks awesome 👍

**Review by atomic-junky — APPROVED** [review]
Sorry for the late review!

**Pull request merged**